In [1]:
import numpy as np
from scipy.stats import fisher_exact, norm

In [2]:
def calc_odds_ratio(a, b, c, d):
    odds_ratio = (a * d) / (b * c)
    p_value = fisher_exact([[a, b], [c, d]])[1]
    se = np.sqrt(1/a + 1/b + 1/c + 1/d)
    z = np.log(odds_ratio) / se
    confint = np.exp(norm.ppf(0.025) * se), np.exp(norm.ppf(0.975) * se)
    return odds_ratio, p_value, confint

In [5]:
#Parent's Occupation
famers_P = 35
famers_B = 75
civil_P = 10
civil_N = 60

odds_ratio, p_value, confint = calc_odds_ratio(famers_P, famers_B, civil_P, civil_N) 

print("The Fisher's Exact Associatio for Parent's Occupation = ")
print("Odds ratio: ", "{:.3f}".format(odds_ratio))
print("P-value: ", "{:.3e}".format(p_value))
print("confint_lower", "{:.3f}".format(confint[0]))
print("confint_upper", "{:.3f}".format(confint[1]))

The Fisher's Exact Associatio for Parent's Occupation = 
Odds ratio:  2.800
P-value:  8.332e-03
confint_lower 0.458
confint_upper 2.183


In [7]:
#Water Activities
bathing_P = 25
bathing_B = 42
washing_P = 10
washing_N = 65

odds_ratio, p_value, confint = calc_odds_ratio(bathing_P, bathing_B, washing_P, washing_N) 

print("The Fisher's Exact Associatio for Water Activities = ")
print("Odds ratio: ", "{:.3f}".format(odds_ratio))
print("P-value: ", "{:.3e}".format(p_value))
print("confint_lower", "{:.3f}".format(confint[0]))
print("confint_upper", "{:.3f}".format(confint[1]))

The Fisher's Exact Associatio for Water Activities = 
Odds ratio:  3.869
P-value:  1.570e-03
confint_lower 0.436
confint_upper 2.293


In [8]:
#Water Supply
well_P = 30
well_B = 50
pond_P = 20
pond_N = 15

odds_ratio, p_value, confint = calc_odds_ratio(well_P, well_B, pond_P, pond_N) 

print("The Fisher's Exact Associatio for Water Supply = ")
print("Odds ratio: ", "{:.3f}".format(odds_ratio))
print("P-value: ", "{:.3e}".format(p_value))
print("confint_lower", "{:.3f}".format(confint[0]))
print("confint_upper", "{:.3f}".format(confint[1]))

The Fisher's Exact Associatio for Water Supply = 
Odds ratio:  0.450
P-value:  6.608e-02
confint_lower 0.446
confint_upper 2.244


In [9]:
#Environmental Sanitation
open_Defacation_P = 20
open_DefacationB = 16
poor_sanitation_P = 10
poor_sanitation_N = 14

odds_ratio, p_value, confint = calc_odds_ratio(open_Defacation_P, open_DefacationB, poor_sanitation_P, poor_sanitation_N) 

print("The Fisher's Exact Associatio for Environmental Sanitation = ")
print("Odds ratio: ", "{:.3f}".format(odds_ratio))
print("P-value: ", "{:.3e}".format(p_value))
print("confint_lower", "{:.3f}".format(confint[0]))
print("confint_upper", "{:.3f}".format(confint[1]))

The Fisher's Exact Associatio for Environmental Sanitation = 
Odds ratio:  1.750
P-value:  4.296e-01
confint_lower 0.352
confint_upper 2.842


**Interpretations** <br>
<br>
The Fisher's exact association test for parent's occupation yielded an odds ratio of 2.800, indicating that the odds of a positive outcome among individuals with parents in farming occupations are approximately 2.8 times higher than those with parents in other occupations. The p-value of 8.332e-03 suggests that this association is statistically significant, providing evidence against the null hypothesis of no association. The 95% confidence interval for the odds ratio ranges from 0.458 to 2.183, indicating the range within which we can be reasonably confident that the true odds ratio lies.
<br>
<br>
For water activities, the odds ratio is 3.869, indicating that the odds of a positive outcome for individuals engaged in bathing activities are about 3.9 times higher than those involved in washing activities. The p-value is 1.570e-03, signifying a statistically significant association. The 95% confidence interval for the odds ratio ranges from 0.436 to 2.293, providing a range of plausible values for the true odds ratio.
<br>
<br>
The Fisher's exact association test for water supply reveals an odds ratio of 0.450, suggesting that the odds of a positive outcome associated with well water are 0.45 times the odds associated with pond water. However, the p-value of 6.608e-02 is greather of 0.05, indicating that the association is not statistically significant.
<br>
<br>
The odds ratio for environmental sanitation is 1.750, implying that the odds of a positive outcome for individuals practicing open defecation are 1.75 times higher than those with poor sanitation practices. However, the p-value of 4.296e-01 is greater than 0.05, suggesting that this association is not statistically significant. The 95% confidence interval for the odds ratio ranges from 0.352 to 2.842.